cmd/spire-server/cli/agent/purge.go

```diff
@@ -0,0 +1,156 @@
+		OutputMask: &types.AgentMask{X509SvidExpiresAt: true},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to list agents: %w", err)
+	}
+
+	agents := resp.GetAgents()
+	expiredAgents := &expiredAgents{Agents: []*expiredAgent{}}
+
+	for _, agent := range agents {
+		id, err := idutil.IDFromProto(agent.Id)
+		if err != nil {
+			return err
+		}
+
+		expirationTime := time.Unix(agent.X509SvidExpiresAt, 0)
+
+		if time.Since(expirationTime) > c.expiredFor {
+			result := &expiredAgent{AgentID: id}
+
+			if !c.dryRun {
+				if _, err := agentClient.DeleteAgent(ctx, &agentv1.DeleteAgentRequest{Id: agent.Id}); err != nil {
+					result.Error = err.Error()
+				} else {
+					result.Deleted = true
+				}
+			}
+			expiredAgents.Agents = append(expiredAgents.Agents, result)
+		}
+	}
+
+	return c.printer.PrintStruct(expiredAgents)
+}
+
+func (c *purgeCommand) AppendFlags(fs *flag.FlagSet) {
+	fs.DurationVar(&c.expiredFor, "expiredFor", 24*time.Hour, "Amount of time that has passed since the agent's SVID has expired. It is used to determine which agents to purge.")
+	fs.BoolVar(&c.dryRun, "dryRun", false, "Indicates that the command will not perform any action, but will print the agents that would be purged.")
+
+	cliprinter.AppendFlagWithCustomPretty(&c.printer, fs, c.env, c.prettyPrintPurgeResult)
+}
+
+type expiredAgents struct {
+	Agents []*expiredAgent `json:"expired_agents"`
+}
+
+type expiredAgent struct {
+	AgentID spiffeid.ID `json:"agent_id"`
+	Deleted bool        `json:"deleted"`
+	Error   string      `json:"error,omitempty"`
+}
+
+func (c *purgeCommand) prettyPrintPurgeResult(env *commoncli.Env, results ...interface{}) error {
+	if expAgents, ok := results[0].([]interface{})[0].(*expiredAgents); ok {
+		if len(expAgents.Agents) == 0 {
+			env.Println("No agents to purge.")
+			return nil
+		}
+
+		msg := fmt.Sprintf("Found %d expired ", len(expAgents.Agents))
+		msg = util.Pluralizer(msg, "agent", "agents", len(expAgents.Agents))
+		env.Printf("%s\n\n", msg)
+
+		if c.dryRun {
+			env.Println("\nAgents that can be purged:")
+			for _, result := range expAgents.Agents {
+				env.Printf("SPIFFE ID         : %s\n", result.AgentID.String())
+			}
+			return nil
+		}
+
+		var agentsNotPurged []*expiredAgent
+		var agentsPurged []*expiredAgent
+
+		for _, result := range expAgents.Agents {
+			if result.Deleted {
+				agentsPurged = append(agentsPurged, result)
+			} else {
+				agentsNotPurged = append(agentsNotPurged, result)
+			}
+		}
+
+		if len(agentsPurged) > 0 {
+			c.printAgentsPurged(agentsPurged)
+		}
+
+		if len(agentsNotPurged) > 0 {
+			c.printAgentsNotPurged(agentsNotPurged)
+		}
+
+		return nil
+	}
+	return cliprinter.ErrInternalCustomPrettyFunc
+}
+
+func (c *purgeCommand) printAgentsNotPurged(agentsNotPurged []*expiredAgent) {
+	c.env.Println("Agents not purged:")
+	for _, result := range agentsNotPurged {
+		c.env.Printf("SPIFFE ID         : %s\n", result.AgentID.String())
+		if result.Error != "" {
```

Comment:
In what case can we have an agent that wasn't purged and also no error string?

Reply:
8901932a41946f3e7644943ed783c1c2f05c5a71